java-source/src/main/java/net/corda/training/api/IOUApi.java

```diff
@@ -0,0 +1,203 @@
+import net.corda.core.internal.FetchDataFlow.Result;
+import net.corda.finance.contracts.GetBalances;
+import net.corda.finance.contracts.asset.Cash;
+import net.corda.finance.flows.CashIssueFlow;
+
+import net.corda.training.flow.IOUIssueFlow;
+import net.corda.training.flow.IOUSettleFlow;
+import net.corda.training.flow.IOUTransferFlow;
+import net.corda.training.flow.SelfIssueCashFlow;
+import net.corda.training.state.IOUState;
+
+import java.util.*;
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x500.style.BCStyle;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import javax.ws.rs.*;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.stream.Collectors;
+
+/**
+ * This API is accessible from /api/iou. The endpoint paths specified below are relative to it.
+ * We've defined a bunch of endpoints to deal with IOUs, cash and the various operations you can perform with them.
+ */
+@Path("iou")
+public class IOUApi{
+    private final CordaRPCOps rpcOps;
+    private final CordaX500Name me;
+
+    private static final Logger logger = LoggerFactory.getLogger(IOUApi.class);
+
+    public IOUApi(CordaRPCOps rpcOps){
+        this.rpcOps = rpcOps;
+        this.me = rpcOps.nodeInfo().getLegalIdentities().get(0).getName();
+    }
+    
+    /** Helpers for filtering the network map cache. */
+    public String toDisplayString(X500Name name){
+        return BCStyle.INSTANCE.toString(name);
+    }
+
+    private boolean isNotary(NodeInfo nodeInfo){
+        if(rpcOps.notaryIdentities()
+            .stream().filter(el -> nodeInfo.isLegalIdentity(el))
+            .collect(Collectors.toList()).size() > 0){
+            return true;
+        }
+        return false;
+    }
+
+    private boolean isMe(NodeInfo nodeInfo){
+        return nodeInfo.getLegalIdentities().get(0).getName().equals(me);
+    }
+
+    private boolean isNetworkMap(NodeInfo nodeInfo){
+        return nodeInfo.getLegalIdentities().get(0).getName().getOrganisation().equals("Network Map Service");
+    }
+    /**
+     * Returns the node's name.
+     */
+    @GET
+    @Path("me")
+    @Produces(MediaType.APPLICATION_JSON)
+    public HashMap<String, String> whoami(){
+        HashMap<String, String> myMap = new HashMap<>();
+        myMap.put("me", me.toString());
+        return myMap;
+    }
+
+    /**
+     * Returns all parties registered with the [NetworkMapService]. These names can be used to look up identities
+     * using the [IdentityService].
+     */
+    @GET
+    @Path("peers")
+    @Produces(MediaType.APPLICATION_JSON)
+    public HashMap<String, List<String>> getPeers(){
+        HashMap<String, List<String>> myMap = new HashMap<>();
+        myMap.put("peers", 
+            rpcOps.networkMapSnapshot().stream().filter(el -> !isNotary(el) && !isMe(el) && !isNetworkMap(el))
+            .map(el -> el.getLegalIdentities().get(0).getName().toString()).collect(Collectors.toList()));
```

Comment:
This expression is incredibly hard to read and reason about. Can we break it down into some number of smaller expressions with meaningful variable names. The compiler will merge them all together anyway. My rule of thumb around complexity is "imagine you have to explain your code to someone else. Do you need to say anything else about the code in order for them to understand it? If you do then its structure needs changing so that this isn't the case, or it needs comments explaining what it's doing"

Reply:
I agree - I've broken down into 3 lines. There's a place in the Contract that will need this rework too. Made other spacing fixes to IOUApi as well.